proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1232 @@
+
+```swift
+extension Span {
+  /// Traps if `position` is not a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - position: an position to validate
+  public boundsCheckPrecondition(_ position: Index)
+
+  /// Traps if `bounds` is not a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - position: a range of positions to validate
+  public boundsCheckPrecondition(_ bounds: Range<Index>)
+}
+```
+
+##### Interoperability with unsafe code:
+
+We provide two functions for interoperability with C or other legacy pointer-taking functions.
+
+```swift
+extension Span {
+  /// Calls a closure with a pointer to the viewed contiguous storage.
+  ///
+  /// The buffer pointer passed as an argument to `body` is valid only
+  /// during the execution of `withUnsafeBufferPointer(_:)`.
+  /// Do not store or return the pointer for later use.
+  ///
+  /// - Parameter body: A closure with an `UnsafeBufferPointer` parameter
+  ///   that points to the viewed contiguous storage. If `body` has
+  ///   a return value, that value is also used as the return value
+  ///   for the `withUnsafeBufferPointer(_:)` method. The closure's
+  ///   parameter is valid only for the duration of its execution.
+  /// - Returns: The return value of the `body` closure parameter.
+  func withUnsafeBufferPointer<Result>(
+    _ body: (_ buffer: UnsafeBufferPointer<Element>) -> Result
+  ) -> Result
+}
+
+extension Span where Element: BitwiseCopyable {
+  /// Calls the given closure with a pointer to the underlying bytes of
+  /// the viewed contiguous storage.
+  ///
+  /// The buffer pointer passed as an argument to `body` is valid only
+  /// during the execution of `withUnsafeBytes(_:)`.
+  /// Do not store or return the pointer for later use.
+  ///
+  /// - Parameter body: A closure with an `UnsafeRawBufferPointer`
+  ///   parameter that points to the viewed contiguous storage.
+  ///   If `body` has a return value, that value is also
+  ///   used as the return value for the `withUnsafeBytes(_:)` method.
+  ///   The closure's parameter is valid only for the duration of
+  ///   its execution.
+  /// - Returns: The return value of the `body` closure parameter.
+  func withUnsafeBytes<Result>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) -> Result
+  ) -> Result
+}
+```
+
+### RawSpan
+
+In addition to `Span<T>`, we propose the addition of `RawSpan`, to represent heterogenously-typed values in contiguous memory. `RawSpan` is similar to `Span<T>`, but represents _untyped_ initialized bytes. `RawSpan` is a specialized type that intends to support parsing and decoding applications, as well as applications where heavily-used code paths require concrete types as much as possible. Its API supports extracting sub-spans, along with the operations `load(as:)` and `loadUnaligned(as:)`.
+
+#### Complete `RawSpan` API:
+
+```swift
+public struct RawSpan: Copyable, ~Escapable {
+  internal var _start: RawSpan.Index
```

Comment:
We don't have `Index`.

Reply:
Found and replaced the last instances of the word `Index`